### ISSUE TITLE
"Dynamic metaclass not supported" false positive on generics

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#11672]
```py
from typing import TypeVar, Generic

T = TypeVar("T")

class FooMetaclass(type, Generic[T]): pass

class Foo(metaclass=FooMetaclass[T]): # error: Dynamic metaclass not supported for "Foo"
    pass
```
https://mypy-play.net/?mypy=master&python=3.10&gist=d4b169a5ece08ed0607b19fff8e9b749

according to [the documentation](https://mypy.readthedocs.io/en/stable/metaclasses.html#gotchas-and-limitations-of-metaclass-support):
> Mypy does not understand dynamically-computed metaclasses, such as `class A(metaclass=f()): ...`

however this isn't dynamically computed, it's gust a generic so mypy should be able to understand it statically

**Comment by sobolevn** [comment]
I am not sure that generic metaclasses is even a thing 🤔 
I've never seen any references in PEPs / docs. Can you please check that it can be used this way?

**Comment by DetachHead** [comment]
pyright supports it and it works at runtime, so i see no reason these two features shouldn't work together for this to emerge.

```py
from typing import TYPE_CHECKING, Generic, TypeVar

T = TypeVar("T")


class FooMetaclass(type, Generic[T]):
    value: T


class Foo(metaclass=FooMetaclass[int]):  # mypy error, no pyright error
    @classmethod
    def foo(cls) -> None:
        if TYPE_CHECKING:
            reveal_type(cls.value)  # pyright says `int`, mypy says `Any`
        cls.value = 1
        print(cls.value)


Foo.foo()  # prints 1
print(Foo.__orig_class__)  # __main__.FooMetaclass[int]
```

**Comment by erictraut** [comment]
Pyright should generate an error in the first example above. The fact that it doesn't is an oversight / bug. `FooMetaclass[T]` shouldn't be allowed because TypeVar `T` doesn't have a defined scope in this case.

Pyright does support the use of a specialized metaclass, as in your second example above. I don't see any reason why this shouldn't work from a type-checking standpoint. That said, it's a use case that isn't well tested, so you may run into issues.

There's an interesting case not covered in either of the above two samples:

```python
from typing import TypeVar, Generic

T = TypeVar("T")

class FooMetaclass(type, Generic[T]):
    pass

class Foo(Generic[T], metaclass=FooMetaclass[T]):
    pass
```

In this variant, TypeVar `T` does have a well-defined scope (since it's bound to class `Foo`), so theoretically, this variant could be valid. I'm not sure it's something we'd want to support though.

@DetachHead, is there a real use case here, or are you just playing around with ideas?

**Comment by last-partizan** [comment]
Those dymanic metaclasses would be really useful for `factory-boy`. I'm trying to add typing support for it, and i've managed to get if work for some cases, but one unsolved issue remains, here's minimal example based on [factory_boy/factory/base.py](https://github.com/FactoryBoy/factory_boy/blob/869e1b0bb104a6881f50117c961031f8a34b91a4/factory/base.py)

```python
from typing import Generic, TypeVar

T = TypeVar("T")


class FactoryMetaClass(type):
    """Factory metaclass for handling ordered declarations."""

    def __call__(cls, **kwargs):
        """Override the default Factory() syntax to call the default strategy.

        Returns an instance of the associated class.
        """
        return cls.create(**kwargs)  # type: ignore


    def __new__(cls, class_name, bases, attrs):
        return super().__new__(cls, class_name, bases, attrs)


class BaseFactory(Generic[T]):
    """Factory base support for sequences, attributes and stubs."""

    @classmethod
    def create(cls, **kwargs) -> T:
        ...


class Factory(BaseFactory[T], metaclass=FactoryMetaClass):
    """Factory base with build and create support.

    This class has the ability to support multiple ORMs by using custom creation
    functions.
    """

# ------------------------------

class Book:
    name: str
    author: str


class BookFactory(Factory[Book]):
    ...


book1 = BookFactory.create()
reveal_type(book1)

book2 = BookFactory()
reveal_type(book2)
```

Both mypy and pyright correctly handling first case, but i have no idea how to get it to return `Book` for second case.
```
mypy basic_example.py
basic_example.py:40:13: note: Revealed type is "basic_example.Book*"
basic_example.py:43:13: note: Revealed type is "basic_example.BookFactory"
```

@erictraut maybe you have idea how to solve this without generic metaclasses?

**Comment by erdnaxeli** [comment]
That's not how FactoryBoy works. It works like this:

```python
class BookFactory(factory.Factory):
    ...

book1 = Bookfactory()
```

You don't instantiate the factory.

**Comment by STofone** [comment]
I need `classproperty` with `generic`.
```python
from functools import cached_property
from typing import Callable, Literal, LiteralString, Type


class SetFunc[T: LiteralString]:
    def __init__(self, id: T):
        self.id = id

    def __call__(self, func: Callable):
        ...
        return func

class META_A[T: LiteralString](type):
    @cached_property
    def set_func(cls) -> Type[SetFunc[T]]:
        return SetFunc

class A[T: LiteralString](metaclass=META_A[T]):
    pass

class B(A[Literal["1", "2", "3"]]):
    pass

# type checking for id does not work
@B.set_func("1")
def func():
    pass
```
